src/vmm/src/rpc_interface.rs

```diff
@@ -441,20 +441,23 @@ impl<'a> PrebootApiController<'a> {
         if load_params.enable_diff_snapshots {
             self.vm_resources.set_track_dirty_pages(true);
         }
-
-        let result = restore_from_snapshot(
+        
+        let result = persist::restore_from_snapshot(
             &self.instance_info,
             &mut self.event_manager,
             self.seccomp_filters,
             load_params,
             VERSION_MAP.clone(),
-        )
-        .and_then(|vmm| {
+        ).and_then(|(vm_resources, vmm)| {
+            println!("{}", serde_json::to_string(&VmmConfig::from(&vm_resources)).unwrap());
+            // TODO: how to deal with lifetime issue here
```

Comment:
The problem here is that you want to take a reference to an owned vmm variable, but the vmm variable is local to the closure. So the reference `self.vm_resources` would live longer than the actual object.

Instead of returning the `VmResources` from `restore_from_snapshot`, you should pass it in there as a mutable reference and modify that variable to contain the device information.